src/game/gui/textinput.c

```diff
@@ -21,56 +23,107 @@ typedef struct {
     surface sur;
     char *buf;
     int max_chars;
+    int pos;
     int bg_enabled;
 } textinput;
 
 static void textinput_render(component *c) {
     textinput *tb = widget_get_obj(c);
-    int chars = strlen(tb->buf);
 
     if(tb->bg_enabled) {
         video_draw(&tb->sur, c->x + (c->w - tb->sur.w) / 2, c->y - 2);
     }
 
     text_mode mode = TEXT_UNSELECTED;
     if(component_is_selected(c)) {
-        if(chars > 0) {
-            mode = TEXT_SELECTED;
-            tb->buf[chars] = '\x7F';
-            tb->buf[chars + 1] = 0;
-            tb->buf[chars] = 0;
+        mode = TEXT_SELECTED;
+        int i = (tb->ticks / 10) % 16;
+        if(i > 8) {
+            i = 16 - i;
+        }
+        tb->tconf.cforeground = 216 + i;
+        int offset = tb->pos * text_char_width(&tb->tconf);
+
+        int start_x = c->x + tb->tconf.padding.left;
+        if(tb->tconf.halign == TEXT_CENTER) {
+            int tmp_s = text_width(&tb->tconf, tb->buf); // Total W minus last spacing
+            int xspace = c->w - tb->tconf.padding.left - tb->tconf.padding.right;
+            start_x += ceilf((xspace - tmp_s) / 2.0f);
+            tb->tconf.halign = TEXT_LEFT;
+            text_render(&tb->tconf, TEXT_DEFAULT, start_x + offset, c->y, c->w, c->h, "\x7F");
+            tb->tconf.halign = TEXT_CENTER;
+        } else {
+            text_render(&tb->tconf, TEXT_DEFAULT, start_x + offset, c->y, c->w, c->h, "\x7F");
         }
     } else if(component_is_disabled(c)) {
         mode = TEXT_DISABLED;
     }
-    if(chars == 0) {
-        mode = TEXT_DISABLED;
-    }
 
     text_render(&tb->tconf, mode, c->x, c->y, c->w, c->h, tb->buf);
 }
 
+static int textinput_action(component *c, int action) {
+    textinput *tb = widget_get_obj(c);
+    DEBUG("action %d", action);
+    switch(action) {
+        case ACT_RIGHT:
+            if(!tb->buf[tb->pos]) {
+                tb->buf[tb->pos] = ' ';
+            }
+            tb->pos = min2(tb->max_chars - 1, tb->pos + 1);
+            if(!tb->buf[tb->pos]) {
+                tb->buf[tb->pos] = ' ';
+            }
+            return 0;
+            break;
+        case ACT_LEFT:
+            tb->pos = max2(0, tb->pos - 1);
+            return 0;
+            break;
+        case ACT_UP:
+            tb->buf[tb->pos] = tb->buf[tb->pos] + 1;
+            if(tb->buf[tb->pos] > 126) {
+                tb->buf[tb->pos] = 32;
+            }
+            return 0;
+            break;
+        case ACT_DOWN:
+            tb->buf[tb->pos] = tb->buf[tb->pos] - 1;
+            if(tb->buf[tb->pos] < 32) {
+                tb->buf[tb->pos] = 126;
+            }
+            return 0;
+            break;
```

Comment:
Seems a bit strange for all these cases to have both return and break, the break would never be reached.

Reply:
I didn't want the code to behave unexpectedly in case someone ever refactored it to remove the return. I think C switch statements should ALWAYS have breaks unless fallthrough is deliberate.